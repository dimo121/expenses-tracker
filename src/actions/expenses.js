import uuid from 'uuid';
import database from '../firebase/firebase';

// Including a database design the action generators will return functions instead of
// objects. Redux does not have usually allow this functionality so we use redux thunk middleware
// to add support for redux to execute the function e.g. firebase push to add object.

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    database.ref(`expenses/${id}`).remove()
    .then(() => {
      dispatch(removeExpense({ id }));
    });
  }
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});


export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    database.ref(`expenses/${id}`).update(updates)
    .then(() => {
      dispatch(editExpense(id,updates));
    });
  }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

//SET_EXPENSES

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});


export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setExpenses(expenses));

    });
  };
};

