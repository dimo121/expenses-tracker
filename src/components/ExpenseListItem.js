import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


//need an expense list item header that takes in filtered expenses and 
//sends back the number of epxenses being shown and their total sum.
//need getExpensesTotal function from expenses-total.js
//should return 0 if no expenses
//should correctly add up a single expense
//should correctly add up multiple expenses
//the header will sit in between the navbar and the expense form
//Need to test these functionalities

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className='list-item' to={`/edit/${id}`}>
    <div>
      <h3 className='list-item__title'>{description}</h3>
      <span className='list-item__sub-title'>{moment(createdAt).format('Do MMMM, YYYY')}</span>
    </div>
    <h3 className='list-item__data'>{numeral(amount / 100).format('$0,0.00')}</h3>
  </Link>
);

export default ExpenseListItem;
