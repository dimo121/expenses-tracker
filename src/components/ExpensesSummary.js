import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import expensesSelector from '../selectors/expenses';
import { getTotalExpenses } from '../selectors/expenses-total';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {

    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses' ;
    const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');

    return (
        <div> 
            <p>Viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</p>
        </div> 
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = expensesSelector(state.expenses, state.filters);
    const total = getTotalExpenses(visibleExpenses);

    return {
        expensesTotal : total,
        expensesCount : visibleExpenses.length
    };
};

export default connect(mapStateToProps)(ExpensesSummary);