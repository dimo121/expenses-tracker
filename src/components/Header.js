import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
  <header className='header'>
    <div className='content-container'>
      <div className='header__content'>
        <Link className='header__title' to='/'>
          <h1>Expensify</h1>
        </Link>
        <Link to="/help" className='button button--link'>Help</Link>
      </div>
    </div>
  </header>
);

export default Header;

//<NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
//<NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>