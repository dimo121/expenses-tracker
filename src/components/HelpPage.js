import React from 'react';
import { Link } from 'react-router-dom';

export default () => {
    return (
        <div>
            <h1>This is the help page</h1><br/>
            <Link to="/">Go home</Link>
        </div>
    )
};