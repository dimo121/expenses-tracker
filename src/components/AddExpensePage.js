import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


export class AddExpensePage extends React.Component{
  
  onSubmit = (expense) => {
    this.props.addExpense(expense);  
    this.props.history.push('/');
  };

  render(){
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }

};



//need a class declaration as the onSubmit is recalculated with every render
//this way the function is refactored in the component class declaration above.

/*
const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        props.onSubmit(expense);
        props.history.push('/');
      }}
    />
  </div>
);

*/

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});


export default connect(undefined,mapDispatchToProps)(AddExpensePage);
