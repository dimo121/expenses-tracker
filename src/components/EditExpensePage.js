import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

//1) Refactor edit expense page to be a class based component this allows us to pull out the inline 
//method definitions and define them before render within the class definition. This saves them 
//redefining each time the component renders.

//2) Setup mapDispatchToProps with two attributes - editExpense and removeExpense and pass the data through

//At this point start the app in development mode to see if it's all working

//3) 3 total test cases:
//  should render EditExpensePage - snapshot
//  should handle editExpense using spies and make sure spies get called
//  should handle removeExpense also using spies similar to the last test in addExpensePage.

export class EditExpensePage extends React.Component{

  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}


const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => ({
  editExpense : (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense : (expense) => dispatch(removeExpense({ id: expense.id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
