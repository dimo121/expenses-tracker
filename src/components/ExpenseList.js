import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export class ExpenseList extends React.Component {
  
  /*
  componentDidMount(){
    this.props.updateCount(this.props.expenses);
    this.props.updateTotal(this.props.expenses);
  }
  */

  render(){
    return (
    <div>
      {
        this.props.expenses.length === 0 ? (
          <p>No expenses</p>
        ) : (
          this.props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )
      }
    </div>
  )};
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

/*
const mapDispatchToProps = (dispatch) => {
  return {
    updateCount : (expenses) => dispatch(setSummaryCount(getExpensesCount(expenses))),
    updateTotal : (expenses) => dispatch(setSummaryTotal(getTotalExpenses(expenses)))
  }
};
*/

export default connect(mapStateToProps)(ExpenseList);
