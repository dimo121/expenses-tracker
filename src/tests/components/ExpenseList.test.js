import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixutres/expenses';

test('should render expenses list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
})

