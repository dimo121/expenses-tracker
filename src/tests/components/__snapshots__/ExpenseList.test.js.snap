// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render expenses list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={4500}
    createdAt={0}
    description="Water bill"
    id="89374edd-2b99-4974-a860-9215c211307f"
    note=""
  />
  <ExpenseListItem
    amount={0}
    createdAt={-345600000}
    description="Gas bill"
    id="fba19938-552e-4ee6-a218-57eba1aea2d6"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={345600000}
    description="Rent"
    id="57be550b-4b83-4bb8-a36d-e3c87338a18b"
    note=""
  />
</div>
`;
